pkg/controller/externalippool/validate.go

```diff
@@ -83,47 +90,132 @@ func (c *ExternalIPPoolController) ValidateExternalIPPool(review *admv1.Admissio
-	} else {
-		if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
-			return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+
+		if gatewayAddr.Is4() {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 32 {
+				return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+			}
+		} else {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
+				return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+			}
 		}
-		mask = net.CIDRMask(int(subnetInfo.PrefixLength), 128)
+		prefix := netip.PrefixFrom(gatewayAddr, int(subnetInfo.PrefixLength)).Masked()
+		subnet = &prefix
 	}
-	subnet := &net.IPNet{
-		IP:   gatewayIP.Mask(mask),
-		Mask: mask,
+
+	// combinedRanges combines both CIDR and start-end style range together mapped to start and end
+	// address of the range. We populate the map with ranges of existing pools and incorporate
+	// the ranges from the current pool as we iterate over them. The map's key is utilized to preserve
+	// the original user-specified input for formatting validation error, if it occurs.
+	combinedRanges := make(map[string][2]netip.Addr)
+	for _, externalIPPool := range existingExternalIPPools {
+		for _, ipRange := range externalIPPool.Spec.IPRanges {
+			var key string
+			var start, end netip.Addr
+
+			if ipRange.CIDR != "" {
+				key = fmt.Sprintf("range [%s] of pool %s", ipRange.CIDR, externalIPPool.Name)
+				cidr, _ := parseIPRangeCIDR(ipRange.CIDR)
+				start, end = utilip.GetStartAndEndOfPrefix(cidr)
+
+			} else {
+				key = fmt.Sprintf("range [%s-%s] of pool %s", ipRange.Start, ipRange.End, externalIPPool.Name)
+				start, end, _ = parseIPRangeStartEnd(ipRange.Start, ipRange.End)
+
+			}
+			combinedRanges[key] = [2]netip.Addr{start, end}
+		}
 	}
+
 	for _, ipRange := range ipRanges {
+		var key string
+		var start, end netip.Addr
+
 		if ipRange.CIDR != "" {
-			_, cidr, err := net.ParseCIDR(ipRange.CIDR)
-			if err != nil {
-				return err.Error(), false
-			}
-			if !ip.IPNetContains(subnet, cidr) {
-				return fmt.Sprintf("cidr %s must be a strict subset of the subnet", ipRange.CIDR), false
+			key = fmt.Sprintf("range [%s]", ipRange.CIDR)
+			cidr, errMsg := parseIPRangeCIDR(ipRange.CIDR)
+			if errMsg != "" {
+				return errMsg, false
 			}
+			start, end = utilip.GetStartAndEndOfPrefix(cidr)
+
 		} else {
-			start := net.ParseIP(ipRange.Start)
-			end := net.ParseIP(ipRange.End)
-			if !subnet.Contains(start) || !subnet.Contains(end) {
-				return fmt.Sprintf("IP range %s-%s must be a strict subset of the subnet", ipRange.Start, ipRange.End), false
+			key = fmt.Sprintf("range [%s-%s]", ipRange.Start, ipRange.End)
+
+			var errMsg string
+			start, end, errMsg = parseIPRangeStartEnd(ipRange.Start, ipRange.End)
+			if errMsg != "" {
+				return errMsg, false
+			}
+
+			// validate if start and end belong to same ip family
+			if start.Is4() && end.Is6() {
```

Comment:
```suggestion
			if start.Is4() != end.Is4() {
```
otherwise it wouldn't detect ipv6 start + ipv4 end